Spigot-Server-Patches/0580-Add-StructureLocateEvent.patch

```diff
@@ -0,0 +1,82 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: dfsek <dfsek@protonmail.com>
+Date: Wed, 16 Sep 2020 01:12:29 -0700
+Subject: [PATCH] Add StructureLocateEvent
+
+
+diff --git a/src/main/java/net/minecraft/server/ChunkGenerator.java b/src/main/java/net/minecraft/server/ChunkGenerator.java
+index 9a6fef215052f9c513b23024968995c97863a453..498a393cdc46cbab0f37a2396a7bba9fbdc6c3e7 100644
+--- a/src/main/java/net/minecraft/server/ChunkGenerator.java
++++ b/src/main/java/net/minecraft/server/ChunkGenerator.java
+@@ -2,6 +2,13 @@ package net.minecraft.server;
+ 
+ import com.google.common.collect.Lists;
+ import com.mojang.serialization.Codec;
++// Paper start
++import org.bukkit.Bukkit;
++import org.bukkit.Location;
++import org.bukkit.StructureType;
++import org.bukkit.event.world.StructureLocateEvent;
++// Paper end
++
+ import java.util.BitSet;
+ import java.util.Iterator;
+ import java.util.List;
+@@ -124,9 +131,34 @@ public abstract class ChunkGenerator {
+ 
+     @Nullable
+     public BlockPosition findNearestMapFeature(WorldServer worldserver, StructureGenerator<?> structuregenerator, BlockPosition blockposition, int i, boolean flag) {
+-        if (!this.b.a(structuregenerator)) {
++        // Paper start
++        org.bukkit.World world = worldserver.getWorld();
++        Location originLocation = new Location(world, blockposition.getX(), blockposition.getY(), blockposition.getZ());
++
++        StructureLocateEvent event = new StructureLocateEvent(world, originLocation, StructureType.getStructureTypes().get(structuregenerator.i()), i, flag);
++        Bukkit.getPluginManager().callEvent(event);
++
++        if(event.isCancelled()) return null;
++
++        // If event call set a final location, skip structure finding and just return set result.
++        Location finalLocation = event.getResult();
++        if(finalLocation != null) {
++            return new BlockPosition(finalLocation.getBlockX(), finalLocation.getBlockY(), finalLocation.getBlockZ());
++        }
++
++        // Get origin location (re)defined by event call.
++        Location newOriginLocation = event.getOrigin();
++        BlockPosition newOriginPosition = new BlockPosition(newOriginLocation.getBlockX(), newOriginLocation.getBlockY(), newOriginLocation.getBlockZ());
++
++        // Get radius and whether to find unexplored structures (re)defined by event call.
++        int radius = event.getRadius();
++        boolean findUnexplored = event.shouldFindUnexplored();
++        StructureGenerator<?> newGenerator = StructureGenerator.a.get(event.getType().getName());
++        // Paper end
++
++        if(! this.b.a(newGenerator)) { // Paper
```

Comment:
Whitespace

Reply:
whitespace issue still exists, but there is no reason why all of these fields need to be modified? just re-assign the existing field, less diff and less likely something will get missed in the future